dartagnan/src/main/java/com/dat3m/dartagnan/wmm/relation/base/RelCrit.java

```diff
@@ -23,59 +30,95 @@ public RelCrit(){
-            // Todo
+            getMaxTupleSet();
         }
         return minTupleSet;
     }
 
     @Override
     public TupleSet getMaxTupleSet(){
-        if(maxTupleSet == null){
-            maxTupleSet = new TupleSet();
-            for(Thread thread : task.getProgram().getThreads()){
-                for(Event lock : thread.getCache().getEvents(FilterBasic.get(Tag.Linux.RCU_LOCK))){
-                    for(Event unlock : thread.getCache().getEvents(FilterBasic.get(Tag.Linux.RCU_UNLOCK))){
-                        if(lock.getCId() < unlock.getCId()){
-                            maxTupleSet.add(new Tuple(lock, unlock));
-                        }
+        if(maxTupleSet != null) {
+            return maxTupleSet;
+        }
+        maxTupleSet = new TupleSet();
+        minTupleSet = new TupleSet();
+        ExecutionAnalysis exec = analysisContext.get(ExecutionAnalysis.class);
+        for (Thread thread : task.getProgram().getThreads()) {
+            // assume order by cId
+            // assume cId describes a topological sorting over the control flow
+            List<Event> locks = reverse(thread.getCache().getEvents(FilterBasic.get(RCU_LOCK)));
+            for (Event unlock : thread.getCache().getEvents(FilterBasic.get(RCU_UNLOCK))) {
+                // iteration order assures that all intermediaries were already iterated
+                for (Event lock : locks) {
+                    if (unlock.getCId() < lock.getCId() ||
+                            exec.areMutuallyExclusive(lock, unlock) ||
+                            concat(minTupleSet.getByFirst(lock).stream().map(Tuple::getSecond),
+                                            minTupleSet.getBySecond(unlock).stream().map(Tuple::getFirst))
+                                    .anyMatch(e -> exec.isImplied(lock, e) || exec.isImplied(unlock, e))) {
+                        continue;
+                    }
+                    boolean noIntermediary = maxTupleSet.getBySecond(unlock).stream()
+                                    .allMatch(t -> exec.areMutuallyExclusive(lock, t.getFirst())) &&
+                            maxTupleSet.getByFirst(lock).stream()
+                                    .allMatch(t -> exec.areMutuallyExclusive(t.getSecond(), unlock));
+                    Tuple tuple = new Tuple(lock, unlock);
+                    maxTupleSet.add(tuple);
+                    if (noIntermediary) {
+                        minTupleSet.add(tuple);
                     }
                 }
             }
-            removeMutuallyExclusiveTuples(maxTupleSet);
         }
         return maxTupleSet;
     }
 
-    // TODO: Not the most efficient implementation
-    // Let's see if we need to keep a reference to a thread in events for anything else, and then optimize this method
+    @Override
+    public void addEncodeTupleSet(TupleSet tuples) {
+        LinkedList<Tuple> queue = new LinkedList<>(intersection(tuples, maxTupleSet));
```

Comment:
You can usse `Deque<Tuple>` with `ArrayDeque<Tuple>` as implementation. Its more natural and likely faster (not that performance matters).

Reply:
Done, thanks!